File: Matrix Calculator/List.java
//Rhea Lingaiah
//rlingaia
//pa3
//11/5/17

public class List {
//----NODE class-------
    private class Node {
        Object data;
        Node next;
        Node prev;

        Node(Object data) {
            this.data = data;
            next = prev = null;
        }

        //to String method
        public String toString() {
            return String.valueOf(data);
        }

        //equals method
        public boolean equals(Object x) {
            boolean eq = false;
            Node that;
            if (x instanceof Node) {
                that = (Node) x;
                eq = (this.data == that.data);
            }
            return eq;
        }
    }
    //--------
    private Node front;
    private Node back;
    private Node cursor;
    private int length;
    private int index;
    List() {
        front = back = cursor = null;
        length = 0;
        index = -1;
    }

    //returns the length
    int length() {
        return length;
    }

    //returns the index of the element if the cursor is not null
    int index() {
        if (cursor != null)
            return index;
        else
            return -1;
    }

    //Returns front element. Pre: length()>0
    Object front() {
        if (length <= 0)
            throw new RuntimeException("called on empty list.");
        return front.data;
    }

    //Returns back element. Pre: length()>0
    Object back() {
        if (length <= 0)
            throw new RuntimeException(" called on empty list.");
        return back.data;
    }


    Object get() {
        //Returns cursor element. Pre: length()>0, index()>=0
        if (length <= 0)
            throw new RuntimeException("called on empty list.");
        if (index < 0)
            throw new RuntimeException("called on invalid index.");
        return cursor.data;
    }


    public boolean equals(List L) {
        // Returns true if and only if this List and L are the same integer sequence
        boolean flag = true; //use flag to break out of the while loop
        Node N = this.front;
        Node M = L.front;
        if (this.length == L.length) {
            while (flag && N != null) { //since the flag is true it goes into the loop
                flag = (N.data == M.data);//check if the data is the same
                N = N.next; //move next in the list
                M = M.next;
            }
            return flag; //if true this will equal true
        } else
            return false;
    }
    //Resets this List to empty state.
    void clear() {
        front = back = cursor = null; //set everything to null
        length = 0;//length gets set to null state
        index = -1; //index gets set to null
    }

    //If List is non-empty, places the cursor under the front element, otherwise does nothing.
    void moveFront() {
        if (length > 0) {
            cursor = front; //move the cursor under the front
            index = 0;
        }
    }

    //If List is non-empty, places the cursor under the back element, otherwise does nothing.
    void moveBack() {
        if (length > 0) {
            cursor = back; //move the cursor under the back
            index = length - 1;
        }
    }

    // If cursor is defined and not at front, moves cursor one step toward
    // front of this List, if cursor is defined and at front, cursor becomes
    // undefined, if cursor is undefined does nothing.
    void movePrev() {
        if (cursor != null && index != 0) {
            cursor = cursor.prev; //if the cursor is not at the front
            index--; //decrement the index
        } else if (cursor != null && index == 0) {
            cursor = null; //if the cursor is at the front
            index = -1;//stays null
        }
    }
    // If cursor is defined and not at back, moves cursor one step toward
    // back of this List, if cursor is defined and at back, cursor becomes
    // undefined, if cursor is undefined does nothing.
    void moveNext() {
        if (cursor != null && index != length - 1) {
            cursor = cursor.next;//move the cursor to next
            index++; //increase the index
        } else if (cursor != null && index == length - 1) {
            cursor = null; //if at the back of the list
            index = -1; //index stays null
        }
    }

    // Insert new element into this List. If List is non-empty,insertion takes place before front element.
    void prepend(Object data) {
        Node newElement = new Node(data);
        if (length <= 0) {
            front = back = newElement; //new element equals the front and back
            length++;
        } else {
            front.prev = newElement;
            newElement.next = front; //the front moves to the next element
            front = newElement;
            length++;//increase the length
            index++; //increase the index

        }

    }

    // Insert new element into this List. If List is non-empty, insertion takes place after back element.
    void append(Object data) {
        Node N;
        if (back == null) { //if the back is null
            back = new Node(data);
            front = back;
            cursor = back;
        } else if (back!=null){
            N = back;
            N.next = new Node(data);
            back = N.next;
            back.prev = N;
        }
        length++;
    }

    // Insert new element before cursor.
    // Pre: length()>0, index()>=0
    void insertBefore(Object data) {
        if (length <= 0 || index < 0)
            throw new RuntimeException("called on an empty list.");
        else {
            if (index == 0)
                prepend(data); //prepend the element if the index is =0
            else {
                Node N = new Node(data);
                Node before = cursor.prev;
                N.prev = before; //insert the node to before
                N.next = cursor; //the cursor
                before.next = N;
                cursor.prev = N;
                length++; //increment the length
                index++;
            }
        }
    }

    // Inserts new element after cursor.
    // Pre: length()>0, index()>=0
    void insertAfter(Object data) {
        if (length <= 0 || index < 0)
            throw new RuntimeException("called on empty list");
        else {
            if (index == length - 1)
                append(data); //append the element
            else {
                Node N = new Node(data);
                Node after = cursor.next;
                N.next = cursor.next;
                N.prev = after;//set the prev=to after
                after.prev = N;
                cursor.next = N;
                length++; //increase the length
            }
        }
    }

    // Deletes the front element. Pre: length()>0
    void deleteFront() {
        if (length < 1)
            throw new RuntimeException("called on an empty List");
        if (cursor == front) {
            cursor = null; //set the front cursor to null
            index = -1;
        }
        front = front.next;
        front.prev = null; //set the front prev null
        length--; //decrement the length and index because deleted an element
        index--;
    }

    // Deletes the back element.
    // Pre: length() > 0
    void deleteBack() {
        if (length < 1) {
            throw new RuntimeException("called on an empty List");
        } else if (cursor == back) { //if the cursor is in the back
            back = back.prev;
            cursor = null; //set to null
            index = -1;
            length--;
        } else {
            back = back.prev;
            back.next = null; //move the next element to null
            length--;
        }
    }

    // Deletes cursor element, making cursor undefined.
    // Pre: length() > 0, index() >= 0
    void delete() {
        if (index < 0 && length < 1)
            throw new RuntimeException("error since length is less than 1 and index is less than 0");
        if (cursor == back)
            deleteBack(); //if the cursor is in the back delete the back
        else if (cursor == front)
            deleteFront(); //delete front if the cursor is in the front
        else {
            cursor.prev.next = cursor.next;
            cursor.next.prev = cursor.prev;
            cursor = null; //set to null
            index = -1;
            length--;
        }
    }

    // Overrides Object's toString method. Returns a String
    public String toString() {
        String str = "";
        Node N;
        for (N = front; N != null; N = N.next) {
            str += N.toString();
            if (N != back) {
                str += " ";
            }
        }
        return str;
    }



    //copys list to a new List
    List copy() {
        List A = new List();
        Node N = this.front;
        while (N != null) {
            A.append(N.data); //COPY ALL THE DATA VALUES INTO THE NEW LIST
            N = N.next;
        }
        return A;
    }
}

File: Depth First Search/List.c
//Rhea Lingaiah
//rlingaia
//PA5

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

typedef struct NodeObj {
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

typedef NodeObj* Node;

typedef struct ListObj {
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;

// Creates a Node with int data, Node prev, and Node next.
Node newNode(int data, Node prev, Node next) {
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->prev = prev;
   N->next = next;
   return(N);
}
// Free memory for the Node.
void freeNode(Node* pN) {
   if(pN != NULL && *pN != NULL) {
      free(*pN);
      *pN = NULL;
   }
}

// Creates a new empty List.
List newList(void){
   List L= malloc(sizeof(ListObj));
   L->front = NULL;
   L->back = NULL;
   L->cursor=NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}

void freeList(List* pL) {
   if (pL != NULL && *pL != NULL) {
      while( length(*pL) > 0 ) {
         deleteBack(*pL);   
      }
      free(*pL);
      *pL = NULL;
   }
}

int length(List L) {
   if(L == NULL) { //check exception
      printf("can't call length on NULL List");
      exit(1);
   }
   return L->length;
}


int index(List L) {
   if(L == NULL) { //check exception
      printf("can't call index on null list");
      exit(1);
   }
   return L->index;
}


int front(List L) {
   if(L == NULL) {//check exception
      printf("can't calling front on null list");
      exit(1);
   }
   else if(L->length < 1) {//check exception
      printf("can't call front on empty List ");
      exit(1);
   }
   return L->front->data;
}


int back(List L) {
   if(L == NULL) {//check exception
      printf("can't call back on NULL List ");
      exit(1);
   }
   else if(L->length < 1) {//check exception
      printf("can't call back on empty List r");
      exit(1);
   }
   return L->back->data;
}

//get data
int get(List L) {
   if(L == NULL) {//check exception
      printf("can't get on NULL List ");
      exit(1);
   }
   else if(L->index < 0||L->length < 1) {//check exception
      printf("can't call get with an undefined index or can't call get on empty List ");
      exit(1);
   }
   return L->cursor->data;
}

//checking if 2 lists are equal
int equals(List A, List B) {
	int flag = 1;
	if(A == NULL || B == NULL) { //check exception
		printf("can't call equals on NULL List");
	}
	Node N = A->front;
	Node M = B->front;
	if(A->length != B->length) {
		return 0;
	}
	while(flag && N != NULL) { //if the flag is true go into loop
		flag=(N->data == M->data); //check if the data is the same
		N=N->next; //move next in the list
		M=M->next;
	}
	return flag; //return the flag
}
//clear the entire list
void clear(List L) {
   if(L == NULL) {//check exception
      printf("can't call clear on NULL List");
      exit(1);
   }
   Node temp = L->front; 
   while(temp != NULL) {
      Node N = temp;
      temp = temp->next; //go through the temp and free each node
      free(N);
   }
   L->front =NULL;
   L->back = NULL;
   L->cursor =NULL;
   L->length =0; //set the length to null
   L->index=-1; //set the index to null
}

//move the cursor to the front of list
void moveFront(List L) {
   if(L == NULL) {//check exception
      printf("can't call moveFront on NULL List ");
      exit(1);
   }
   if(L->length > 0) {
      L->cursor = L->front; //move the cursor the front
      L->index=0;
   }
}
//move the cursor the back of list
void moveBack(List L) {
   if(L == NULL) {//check exception
      printf("can't call moveBack on NULL List");
      exit(1);
   }
   if(L->length > 0) {
      L->cursor = L->back; //move the cursor back
      L->index = L->length - 1;
   }
}
//move the cursor prev
void movePrev(List L) {
   if(L == NULL) { //check exception
      printf("can't call movePrev on null List");
      exit(1);
   }
   if(L->cursor != NULL && L->index != 0) {
      L->cursor = L->cursor->prev; //if the cursor isn't the front
      L->index--; //decrement the index
   }
   else if(L->cursor != NULL && L->index == 0) {
      L->cursor = NULL; //if the cursor is at the front
      L->index = -1; //index is null
   }
}
//move the cursor next
void moveNext(List L) {
   if(L == NULL) { //check exception
      printf("can't call moveNext on null List");
      exit(1);
   }
   if(L->cursor != NULL && L->index != L->length - 1) {
      L->cursor = L->cursor->next; //move the cursor to next
      ++L->index; //increase the index
   }
   else if(L->cursor != NULL && L->index == L->length - 1) {
      L->cursor = NULL;  //if at the back of the list
      L->index = -1; //index stays null
   }
}
//inserts the node before
void prepend(List L, int data) {
   if(L == NULL) { //list exception check
      printf("can't call prepend on null List");
      exit(1);
   }
   Node N = newNode(data, NULL, L->front);
   if(L->front == NULL)
      L->back = N;    //set the back
   else if(L->front!=NULL){
      L->front->prev = N;
   }
   L->front = N;
   L->index++;       //increase the index
   L->length++;      //increase the index
}
//inserts the node after
void append(List L, int data){
   if(L==NULL){
      printf("can't call append on null list");
      exit(1);
   }
	Node N = newNode(data,L->back, NULL);
	if(L->length == 0){ 
		L->front = L->back = N; //if the length is 0 then set the front and back to N
	}else{
		Node P;
		P = L->back;
		L->back->next= N; //set the back to next
		L->back=N;
		N->prev =P;
		P->next =N;
		}
	L->length++;     //increment the length
	}
//inserts before the cursor
void insertBefore(List L, int data) {
   if(L == NULL) { //exception check
      printf("can't call insertBefore on null List");
      exit(1);
   }
   else if(L->index < 0||L->length < 1) { //exception check
      printf("can't call insertBefore on null index or can't call insertBefore called on an empty List");
      exit(1);
   }
   Node N = newNode(data, L->cursor->prev, L->cursor);
   if(L->cursor->prev != NULL)
      L->cursor->prev->next = N;  //insert the node before
   else{
      L->front = N;
   }
   L->cursor->prev = N; //insert the node before
   L->length++;      //increment the length
   L->index++;       //increment the index
}
//inserts after the cursor
void insertAfter(List L, int data) {
   if(L==NULL) { //exception check
      printf("cant call insertafter on Null list");
      exit(1);
   }
   else if(L->length<1||L->index<0) { //exception check
      printf("can't call insertAfter when length is 0 or cant call insertafter on null index");
      exit(1);
   }
   Node N = newNode(data, L->cursor, L->cursor->next);
   if(L->cursor->next != NULL)   //check if the next is null
      L->cursor->next->prev=N;  
   else if(L->cursor->next == NULL){
      L->back=N;
   }
   L->cursor->next = N;
   L->length++; //increase the length
}
//deletes the front element
void deleteFront(List L){
	if(L->length < 1){//exception check
		printf( "cant call deleteFront when the length is less than 1");
		exit(1);
	}
	Node N=L->front;
	if(length(L)>1){	
		L->front=L->front->next;	
		L->front->prev=NULL; //set the front prev null
		L->index--;
	}else{
		L->front = NULL;
		L->back = NULL;
		L->index = -1;
		}
	L->length--; //decrement the length and index because deleted an element
	freeNode(&N);
	}
//deletes the back element
void deleteBack(List L){
   if(length(L)<1){ //exception to check deletes
   	printf("deleteBack can't be called if the length is less than 1");
   	exit(1);
   }
   Node N = L->back;
   if(length(L) == 1){
      L->front = L->back = NULL;
      L->index = -1;
   }
   else if( index(L)!=length(L)-1&&length(L)>1){ //if the cursor is in the back
      L->back=L->back->prev;
      L->back->next=NULL; //set to null
   }
   else if(length(L)>1 && index(L)==length(L)-1){
   	L->back = L->back->prev;
   	L->back->next = NULL; //move the next element to null
   	L->index = -1;
   }
   L->length--; //decrement the length
   freeNode(&N); //free
}

void delete(List L) {
	if(L == NULL) { //exception to check deletes
		printf("can't call delete on a null list");
		exit(1);
		}
	else if(index(L) == -1) { //exception to check deletes
		printf("delete on undefined list that is null");
		exit(1);
	} else if (length(L) == 0) {
		printf("delete on empty List");
		exit(1);
	}
	if (L->cursor == L->back) {
      deleteBack(L); //if the cursor is in the back delete the back
   } else if(L->cursor == L->front) {
	  deleteFront(L); //if the cursor is in the front delete the front
	}
   else{
   	Node left = L->cursor->prev;
   	Node right = L->cursor->next;
   	Node N = L->cursor;
   	left->next = right;
   	right->prev = left;
   	L->cursor = NULL; //set the cursor to null
   	L->index = -1; //set the index to null
   	L->length--; //dec the length since removing a node
   	freeNode(&N); //free
	}
}

//printing the list of nodes out
void printList(FILE* out, List L) {
    if(L == NULL) {
        printf("can't call printList on null List");
        exit(1);
    }
    Node N = L->front;
    int i = 0;
    while(N != NULL) {
      if(i != 0)
        	fprintf(out, " %d", N->data); //print the data of the node
      else 
        	fprintf(out, "%d", N->data); //print the data of the node
      N = N->next; //go onto the next node
      i++;
    }
}
//copy into the new list
List copyList(List L) {
   List A = newList();
   Node N = L->front;
   while(N != NULL) {
      append(A, N->data); //copy the values into the new list
      N = N->next;
   }
   return A;
}
